offscreen/scripts/offscreen.ts

```diff
@@ -9,41 +10,63 @@ import initLedger from './ledger';
 import initTrezor from './trezor';
 import initLattice from './lattice';
 
-initLedger();
-initTrezor();
-initLattice();
-
 /**
  * Initialize a post message stream with the parent window that is initialized
  * in the metamask-controller (background/serivce worker) process. This will be
  * utilized by snaps for communication with snaps running in the offscreen
  * document.
  */
-const parentStream = new BrowserRuntimePostMessageStream({
-  name: 'child',
-  target: 'parent',
-});
-
-ProxySnapExecutor.initialize(parentStream, './snaps/index.html');
-
-if (process.env.IN_TEST) {
-  chrome.runtime.onMessage.addListener((message) => {
-    if (
-      message &&
-      typeof message === 'object' &&
-      message.event === OffscreenCommunicationEvents.metamaskBackgroundReady &&
-      message.target === OffscreenCommunicationTarget.extension
-    ) {
-      window.document?.documentElement?.classList?.add('controller-loaded');
-    }
+function initializePostMessageStream() {
+  const parentStream = new BrowserRuntimePostMessageStream({
+    name: 'child',
+    target: 'parent',
   });
+
+  ProxySnapExecutor.initialize(parentStream, './snaps/index.html');
 }
 
-chrome.runtime.sendMessage({
-  target: OffscreenCommunicationTarget.extensionMain,
-  isBooted: true,
+/**
+ * Initialize the ledger, trezor, and lattice keyring connections, and the
+ * post message stream for the Snaps environment.
+ */
+async function init(): Promise<void> {
+  initializePostMessageStream();
+  initTrezor();
+  initLattice();
 
-  // This message is being sent from the Offscreen Document to the Service Worker.
-  // The Service Worker has no way to query `navigator.webdriver`, so we send it here.
-  webdriverPresent: navigator.webdriver === true,
+  try {
+    const ledgerInitTimeout = new Promise((_, reject) => {
+      setTimeout(() => {
+        reject(new Error('Ledger initialization timed out'));
+      }, OFFSCREEN_LEDGER_INIT_TIMEOUT);
+    });
+    await Promise.race([initLedger(), ledgerInitTimeout]);
+  } catch (error) {
+    console.error('Ledger initialization failed:', error);
+  }
+}
+
+init().then(() => {
+  if (process.env.IN_TEST) {
+    chrome.runtime.onMessage.addListener((message) => {
+      if (
+        message &&
+        typeof message === 'object' &&
```

Comment:
nit: This is a stricter check (although it shouldn't make a significant difference here).
```suggestion
        isObject(message) &&
```

Reply:
Refactored in e764bdf3308bdd105ddbef611a2188fb57e38866 along with 

```typescript
import { isObject } from '@metamask/utils';
```